beacon-chain/powchain/deposit.go

```diff
@@ -1,82 +1,34 @@
 )
 
-// processDeposit is a copy of the core function of the same name which includes some optimizations
-// and removes the requirement to pass in beacon state. This is for determining genesis validators.
-func (s *Service) processDeposit(
-	eth1Data *ethpb.Eth1Data,
-	deposit *ethpb.Deposit,
-) error {
-	if err := verifyDeposit(eth1Data, deposit); err != nil {
-		return errors.Wrapf(err, "could not verify deposit from %#x", bytesutil.Trunc(deposit.Data.PublicKey))
+func (s *Service) processDeposit(eth1Data *ethpb.Eth1Data, deposit *ethpb.Deposit) error {
+	var err error
+	valIndexMap := stateutils.ValidatorIndexMap(s.preGenesisState)
+	s.preGenesisState.Eth1Data = eth1Data
+	s.preGenesisState, err = blocks.ProcessDeposit(s.preGenesisState, deposit, valIndexMap)
+	if err != nil {
+		return errors.Wrap(err, "could not process deposit")
 	}
-	pubKey := bytesutil.ToBytes48(deposit.Data.PublicKey)
-	amount := deposit.Data.Amount
-	currBal, ok := s.depositedPubkeys[pubKey]
+	pubkey := deposit.Data.PublicKey
+	index, ok := valIndexMap[bytesutil.ToBytes48(pubkey)]
 	if !ok {
-		pub, err := bls.PublicKeyFromBytes(pubKey[:])
-		if err != nil {
-			return errors.Wrap(err, "could not deserialize validator public key")
-		}
-		domain := bls.ComputeDomain(params.BeaconConfig().DomainDeposit)
-		sig, err := bls.SignatureFromBytes(deposit.Data.Signature)
-		if err != nil {
-			return errors.Wrap(err, "could not convert bytes to signature")
-		}
-		root, err := ssz.SigningRoot(deposit.Data)
-		if err != nil {
-			return errors.Wrap(err, "could not sign root for deposit data")
-		}
-		if !sig.Verify(root[:], pub, domain) {
-			return fmt.Errorf("deposit signature did not verify")
-		}
-		s.depositedPubkeys[pubKey] = amount
-
-		if amount >= params.BeaconConfig().MaxEffectiveBalance {
-			s.activeValidatorCount++
-		}
-	} else {
-		newBal := currBal + amount
-		s.depositedPubkeys[pubKey] = newBal
-		// exit if the validator is already an active validator previously
-		if currBal >= params.BeaconConfig().MaxEffectiveBalance {
-			return nil
-		}
-		if newBal >= params.BeaconConfig().MaxEffectiveBalance {
-			s.activeValidatorCount++
-		}
-	}
-	return nil
-}
-
-func verifyDeposit(eth1Data *ethpb.Eth1Data, deposit *ethpb.Deposit) error {
-	// Verify Merkle proof of deposit and deposit trie root.
-	receiptRoot := eth1Data.DepositRoot
-	leaf, err := ssz.HashTreeRoot(deposit.Data)
-	if err != nil {
-		return errors.Wrap(err, "could not tree hash deposit data")
+		return nil
 	}
-	if ok := trieutil.VerifyMerkleProof(
-		receiptRoot,
-		leaf[:],
-		int(eth1Data.DepositCount-1),
-		deposit.Proof,
-	); !ok {
-		return fmt.Errorf(
-			"deposit merkle branch of deposit root did not verify for root: %#x",
-			receiptRoot,
-		)
+	balance := s.preGenesisState.Balances[index]
+	s.preGenesisState.Validators[index].EffectiveBalance = mathutil.Min(balance-balance%params.BeaconConfig().EffectiveBalanceIncrement, params.BeaconConfig().MaxEffectiveBalance)
```

Comment:
this is a very long expression. Can these be put in perhaps some small helper funcs in this file? Or maybe some of  of the parts of this expression can be broken up into variables, we also need comprehensive tests for this particular few lines